### PR TITLE
Support fetch executed tensor data from server

### LINKED ISSUE #17: [BUG][DISTRIBUTED] client will hang if submit the same graph again
**Describe the bug**
if we resubmit the same tensor by session, the client will hang, at the same time, i found scheduler raise an  error `ActorAlreadyExist`, which i think can be solved by prune existed operand.

**To Reproduce**
I creat a distributed environment with 1 worker.

```
In [1]: import mars.tensor as mt

In [2]: from mars.session import new_session

In [3]: session = new_session('http://0.0.0.0:33335')

In [4]: a = mt.ones((10, 10)) + 1

In [5]: session.run(a)
Out[5]: 
array([[2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.]])

In [6]: session.run(a)

```
the client hang when execute line 6, and scheduler log shows:
```
Creating operand actors for graph 93ba7372-e748-40ab-af14-6c250f845ba4
Traceback (most recent call last):
  File "src/gevent/greenlet.py", line 716, in gevent._greenlet.Greenlet.run
  File "mars/actors/pool/gevent_pool.pyx", line 88, in mars.actors.pool.gevent_pool.ActorExecutionContext.fire_run
  File "mars/actors/pool/gevent_pool.pyx", line 91, in mars.actors.pool.gevent_pool.ActorExecutionContext.fire_run
  File "mars/actors/pool/gevent_pool.pyx", line 102, in mars.actors.pool.gevent_pool.ActorExecutionContext.fire_run
  File "mars/actors/pool/gevent_pool.pyx", line 96, in mars.actors.pool.gevent_pool.ActorExecutionContext.fire_run
  File "mars/actors/core.pyx", line 108, in mars.actors.core.FunctionActor.on_receive
  File "mars/actors/core.pyx", line 110, in mars.actors.core.FunctionActor.on_receive
  File "/Users/hekaisheng/Documents/mars/mars/scheduler/graph.py", line 156, in execute_graph
    self.create_operand_actors()
  File "/Users/hekaisheng/Documents/mars/mars/scheduler/graph.py", line 569, in create_operand_actors
    [future.result() for future in itertools.chain(six.itervalues(op_refs), [write_future])]
  File "/Users/hekaisheng/Documents/mars/mars/scheduler/graph.py", line 569, in <listcomp>
    [future.result() for future in itertools.chain(six.itervalues(op_refs), [write_future])]
  File "src/gevent/event.py", line 457, in gevent._event.AsyncResult.result
  File "src/gevent/event.py", line 381, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 399, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 379, in gevent._event.AsyncResult._raise_exception
  File "/Users/hekaisheng/miniconda3/lib/python3.6/site-packages/gevent/_compat.py", line 47, in reraise
    raise value.with_traceback(tb)
  File "mars/actors/pool/gevent_pool.pyx", line 758, in mars.actors.pool.gevent_pool.Communicator._create_local_actor
  File "mars/actors/pool/gevent_pool.pyx", line 237, in mars.actors.pool.gevent_pool.LocalActorPool.create_actor
mars.actors.errors.ActorAlreadyExist: Actor s:operator$2724de06-fc4a-11e8-8d24-47a7eeef77d5$c84e3ef10df7fabaf2dc028149609284 already exist, cannot create

```

### PR DESCRIPTION
**Pull request by NJUPole**
## What do these changes do?

Add `session.fetch` method and check if tensor has been submitted in `session.run`. For `LocalClusterSession` and `web.Session`, they can fetch executed tensor from workers and submit other tensors to cluster. For `LocalSession`, fetch is not available, it will run all input tensors whether they have run or not.

## Related issue number
fixes #17 from the client side, other related fixes may be done with #28

### CODE REVIEW

**Comment by chineking** [comment]
Wonder why local session cannot fetch? `LocalSession` hold an executor which have all the executed data.

**Comment by NJUPole** [comment]
> Wonder why local session cannot fetch? `LocalSession` hold an executor which have all the executed data.

okay...`Executor` hold data in `_chunk_result`, it can provide interface for fetch.

**Comment by chineking** [comment]
Yeah, add an interface to Executor to supply `fetch` ability.

File: mars/web/session.py [hunk 1]
```diff
@@ -21,8 +21,7 @@
 
 from ..compat import six, TimeoutError
 from ..serialize import dataserializer
-from ..errors import ResponseMalformed, ExecutionInterrupted, ExecutionFailed, \
-    ExecutionStateUnknown, ExecutionNotStopped
+from ..errors import *
```
Comment (wjsi) [inline mars/web/session.py]: Flake 8 discourages using import star in code.

File: mars/deploy/local/session.py [hunk 2]
```diff
@@ -81,6 +81,11 @@ def run(self, *tensors, **kw):
 
         return data_list
 
+    def fetch(self, key):
+        graph_key = self._tensor_to_graph[key]
+        resp = self._api.fetch_data(self._session_id, graph_key, key)
+        return dataserializer.loads(resp)
```
Comment (wjsi) [inline mars/deploy/local/session.py]: I think the MarsAPI class should handle data serialization / deserialization itself.

File: mars/session.py [hunk 3]
```diff
@@ -39,6 +39,9 @@ def run(self, *tensors, **kw):
             raise RuntimeError('Session has closed')
         return self._executor.execute_tensors(tensors, **kw)
 
+    def fetch(self, key):
+        raise ValueError('Local session cannot fetch data')
```
Comment (wjsi) [inline mars/session.py]: NotImplementedError ?

Comment (chineking) [reply mars/session.py]: This should be implemented.

File: mars/session.py [hunk 4]
```diff
@@ -39,6 +39,24 @@ def run(self, *tensors, **kw):
...
+        if len(tensor.chunks) == 1:
+            return self._executor.chunk_result[tensor.chunks[0].key]
+
+        chunks = []
+        for c in tensor.chunks:
+            op = TensorFetchChunk(dtype=c.dtype, to_fetch_key=c.key)
+            chunk = op.new_chunk(None, c.shape, index=c.index, _key=c.key)
+            chunks.append(chunk)
+
+        new_op = TensorFetchChunk(dtype=tensor.dtype, to_fetch_key=tensor.key)
```
Comment (chineking) [inline mars/session.py]: No need to generate a new tensor op, just `new_op=tensor.op.copy()` seems fine.

File: mars/session.py [hunk 5]
```diff
@@ -39,6 +39,24 @@ def run(self, *tensors, **kw):
             raise RuntimeError('Session has closed')
         return self._executor.execute_tensors(tensors, **kw)
 
+    def fetch(self, tensor):
+        from .tensor.expressions.datasource import TensorFetchChunk
```
Comment (chineking) [inline mars/session.py]: Check if the tensor is executed, if not raise an error.

Comment (NJUPole) [reply mars/session.py]: I think it should be checked in `Mars.Session.fetch`

File: mars/session.py [hunk 6]
```diff
@@ -173,8 +176,8 @@ def __enter__(self):
         self._sess.__enter__()
         return self
 
-    def __exit__(self, exc_type, exc_val, exc_tb):
-        self._sess.__exit__(exc_type, exc_val, exc_tb)
+    def __exit__(self, *_):
```
Comment (chineking) [inline mars/session.py]: Better to keep original `__exit__`, and redefine `close` method.

**Pull request merged**